hammer/workers.go

```diff
@@ -0,0 +1,203 @@
+// Copyright 2024 The Tessera authors. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package main
+
+import (
+	"context"
+	"encoding/base64"
+	"errors"
+	"fmt"
+	"math/rand"
+
+	"github.com/transparency-dev/trillian-tessera/client"
+	"k8s.io/klog/v2"
+)
+
+// LeafWriter is the signature of a function which can write arbitrary data to a log.
+// The data to be written is provided, and the implementation must return the sequence
+// number at which this data will be found in the log, or an error.
+type LeafWriter func(ctx context.Context, data []byte) (uint64, error)
+
+// NewLeafReader creates a LeafReader.
+// The next function provides a strategy for which leaves will be read.
+// Custom implementations can be passed, or use RandomNextLeaf or MonotonicallyIncreasingNextLeaf.
+func NewLeafReader(tracker *client.LogStateTracker, f client.Fetcher, next func(uint64) uint64, throttle <-chan bool, errchan chan<- error) *LeafReader {
+	return &LeafReader{
+		tracker:  tracker,
+		f:        f,
+		next:     next,
+		throttle: throttle,
+		errchan:  errchan,
+	}
+}
+
+// LeafReader reads leaves from the tree.
+// This class is not thread safe.
+type LeafReader struct {
+	tracker  *client.LogStateTracker
+	f        client.Fetcher
+	next     func(uint64) uint64
+	throttle <-chan bool
+	errchan  chan<- error
```

Comment:
nit: `errChan` to be consistent with caps in hammer.go

Reply:
Done.